hydra_base/lib/storage/hdfstorageadapter.py

```diff
@@ -290,13 +227,102 @@ def retrieve_s3_file(self, url, **kwargs):
+        for p in glob.iglob(pattern, recursive=True):
+            if not os.path.isfile(p):
+                continue
+            files[p] = os.stat(p).st_size
+        return files
+
+    def purge_local_file(self, filename):
+        """
+          This prevents directory traversal by:
+            - relative path components
+            - ~user path components
+            - $ENV_VAR components
+            - paths containing hard or symbolic links
+
+          A valid target file must be all of:
+            - a real absolute filesystem path
+            - a subtree of the filestore
+            - not a directory
+            - not a link
+            - not a device file or pipe
+            - owned by the Hydra user
+
+          In addition, the filestore_path may not be:
+            - undefined
+            - the root filesystem
+            - the root of any mount point
+
+          ValueError is raised if any of these conditions
+          are not met.
+        """
+        real_fsp = os.path.realpath(self.filestore_path)
+        if not self.filestore_path or real_fsp == '/' or os.path.ismount(real_fsp):
+            raise ValueError(f"Invalid filestore configuration value '{self.filestore_path}'")
+
+        expanded = os.path.expandvars(filename)
+        if expanded != filename:
+            raise ValueError(f"Invalid path '{filename}': Arguments may not contain variables")
+        target = os.path.realpath(expanded)
+        if os.path.commonprefix([target, self.filestore_path]) != self.filestore_path:
+            raise ValueError(f"Invalid path '{filename}': Only filestore files may be purged")
+
+        if not os.path.exists(target):
+            raise ValueError(f"Invalid path '{filename}': File does not exist")
+
+        # Tests for directories, device files and pipes, and existence again
+        if not os.path.isfile(target):
+            raise ValueError(f"Invalid path '{filename}': Only regular files may be purged")
+
+        if os.getuid() != os.stat(target).st_uid:
+            raise ValueError(f"Invalid path '{filename}': File is not owned by "
+                             f"user {os.getlogin()} ({os.getuid()})")
+        try:
+            os.unlink(target)
+        except OSError as oe:
+            raise ValueError(f"Invalid path '{filename}': Unable to purge file") from oe
+
+        return target
 
-def nscale(ts):
-    """
-      Transforms integers representing nanoseconds past the epoch
-      into instances of datetime.timestamp
-    """
-    return datetime.fromtimestamp(ts/1e9)
+    def make_group_reader(self, url, groupname):
+        hf = self.open_hdf_url(url)
+        if not groupname:
+            # Use first group in file
+            try:
+                groupname = [*hf.keys()][0]
+            except IndexError as ie:
+                raise ValueError(f"Data source {url} contains no groups") from ie
+        Reader = self.get_group_reader(url, groupname)
+        return Reader(hf, groupname)
```

Comment:
The `group_reader_map` is an index of *types* which is created automatically by the `__init_subclass__` method of the `GroupReader` ABC.  This function "validates" subtypes by requiring them to define a `subclass_type_key` attr which indicates the `pandas_type` for which they are a reader, i.e. the overridden abstract methods each class implements are specific to layout of data in the HDF group which has the same `pandas_type` value.

I've expanded the comment in `GroupReader` to make it clear this happens automatically.  In the event that a different internal HDF format is encountered, we just need to make a new subclass with implementations of the abstract methods which handle that format, and it will automatically be available to the storage adapter layer as a reader,